src/adapters/encounter-diagnosis-adapter.ts

```diff
@@ -0,0 +1,84 @@
+import { type FormContextProps } from '../provider/form-provider';
+import { type OpenmrsEncounter, type FormField } from '../types';
+import { clearSubmission, gracefullySetSubmission } from '../utils/common-utils';
+
+export let assignedDiagnosesIds: string[] = [];
+
+export const EncounterDiagnosisAdapter: FormFieldValueAdapter = {
+  transformFieldValue: function (field: FormField, value: any, context: FormContextProps) {
+    if (context.sessionMode == 'edit' && field.meta?.previousValue?.uuid) {
+      return editDiagnosis(value, field, context.patient.id);
+    }
+    const newValue = constructNewDiagnosis(value, field, context.patient.id);
+    gracefullySetSubmission(field, newValue, null);
+    return newValue;
+  },
+  getInitialValue: function (
+    field: FormField,
+    sourceObject: OpenmrsResource,
+    context: FormProcessorContextProps,
+  ): Promise<any> {
+    const encounter = sourceObject ?? (context.domainObjectValue as OpenmrsEncounter);
+    const matchedDiagnosis = encounter.diagnoses.find(
+      (diagnosis) => diagnosis.formFieldPath === `rfe-forms-${field.id}`,
+    );
+
+    if (matchedDiagnosis) {
+      field.meta = { ...(field.meta || {}), previousValue: matchedDiagnosis };
+      if (!assignedDiagnosesIds.includes(matchedDiagnosis.diagnosis?.coded?.uuid)) {
+        assignedDiagnosesIds.push(matchedDiagnosis.diagnosis?.coded?.uuid);
+      }
+      return matchedDiagnosis.diagnosis?.coded.uuid;
+    }
+    return null;
+  },
+  getPreviousValue: function (
+    field: FormField,
+    sourceObject: OpenmrsResource,
+    context: FormProcessorContextProps,
+  ): Promise<any> {
+    return null;
+  },
+  getDisplayValue: (field: FormField, value: any) => {
+    return field.questionOptions.answers?.find((option) => option.concept == value)?.label || value;
+  },
+  tearDown: function (): void {
+    assignedDiagnosesIds = [];
+  },
+};
+
+const constructNewDiagnosis = (value: any, field: FormField, patientUuid: string, uuid?: string) => {
+  if (!value) {
+    return null;
+  }
+  const diagnosis: DiagnosisPayload = {
+    patient: patientUuid,
+    condition: null,
+    diagnosis: {
+      coded: value,
+    },
+    certainty: field.questionOptions?.diagnosis?.isConfirmed ? 'CONFIRMED' : 'PROVISIONAL',
+    rank: field.questionOptions.diagnosis?.rank ?? 1, // rank 1 denotes a diagnosis is primary, else secondary
+    formFieldPath: `rfe-forms-${field.id}`,
+    formFieldNamespace: 'rfe-forms',
+  };
+
+  if (uuid && uuid.trim() !== '') {
+    diagnosis.uuid = uuid;
+  }
+  return diagnosis;
+};
+
+function editDiagnosis(newEncounterDiagnosis: any, field: FormField, patientUuid: string) {
+  if (newEncounterDiagnosis === field.meta.previousValue?.diagnosis?.coded?.uuid) {
+    clearSubmission(field);
+    return null;
+  }
+  const oldDiagnosis = field.meta.initialValue?.omrsObject as OpenmrsResource;
+
+  const newValue = constructNewDiagnosis(newEncounterDiagnosis, field, patientUuid, oldDiagnosis?.uuid);
```

Comment:
What happens if `newEncounterDiagnosis` is null?

Reply:
If its null, there is a block that handles that in `constructNewDiagnosis`

Reply:
The factory function that constructs a new diagnosis would return null! Considering that we are in the context of editing an existing diagnosis, wouldn't this mean that the user deleted the field hence we should void the old diagnosis? 

```ts
const constructNewDiagnosis = (value: any, field: FormField, patientUuid: string, uuid?: string) => {
  if (!value) {
    return null;
  }

```